src/lib/pubkey/sphincsplus/sphincsplus_common/sp_hash.cpp

```diff
@@ -0,0 +1,95 @@
+/*
+ * SPHINCS+ Hashes
+ * (C) 2023 Jack Lloyd
+ *     2023 Fabian Albert, René Meusel, Amos Treiber - Rohde & Schwarz Cybersecurity
+ *
+ * Botan is released under the Simplified BSD License (see license.txt)
+ **/
+
+#include <botan/internal/sp_hash.h>
+
+#include <botan/internal/stl_util.h>
+
+#include <botan/assert.h>
+#include <botan/exceptn.h>
+#include <botan/hash.h>
+#include <botan/sp_parameters.h>
+
+#if defined(BOTAN_HAS_SPHINCS_PLUS_WITH_SHAKE)
+   #include <botan/internal/sp_hash_shake.h>
+#endif
+
+#if defined(BOTAN_HAS_SPHINCS_PLUS_WITH_SHA2)
+   #include <botan/internal/sp_hash_sha2.h>
+#endif
+
+#include <memory>
+
+namespace Botan {
+
+Sphincs_Hash_Functions::Sphincs_Hash_Functions(const Sphincs_Parameters& sphincs_params,
+                                               const SphincsPublicSeed& pub_seed) :
+      m_sphincs_params(sphincs_params), m_pub_seed(pub_seed) {}
+
+std::unique_ptr<Sphincs_Hash_Functions> Sphincs_Hash_Functions::create(
+   const Sphincs_Parameters& sphincs_params, [[maybe_unused]] const SphincsPublicSeed& pub_seed) {
+   switch(sphincs_params.hash_type()) {
+      case Sphincs_Hash_Type::Sha256:
+#if defined(BOTAN_HAS_SPHINCS_PLUS_WITH_SHA2)
+         return std::make_unique<Sphincs_Hash_Functions_Sha2>(sphincs_params, pub_seed);
+#else
+         throw Not_Implemented("SPHINCS+ with SHA-256 is not available in this build");
+#endif
+      case Sphincs_Hash_Type::Haraka:
+         throw Not_Implemented("Haraka is not yet implemented");
+      case Sphincs_Hash_Type::Shake256:
+#if defined(BOTAN_HAS_SPHINCS_PLUS_WITH_SHAKE)
+         return std::make_unique<Sphincs_Hash_Functions_Shake>(sphincs_params, pub_seed);
+#else
+         throw Not_Implemented("SHAKE is not available in this build");
+#endif
+   }
+
+   Botan::unreachable();
```

Comment:
This whole function is pretty hard to understand, and it's not entirely obvious at first glance that the unreachable really is unreachable. Also if we are passed an invalid `enum` value somehow, we have UB.

I'd suggest something like

```
switch(sphinx_params.hash_type()) {
   #if defined(BOTAN_HAS_SPHINCS_PLUS_WITH_SHA2)
   case Sphincs_Hash_Type::Sha256:
            return std::make_unique<Sphincs_Hash_Functions_Sha2>(sphincs_params, pub_seed);
   #endif 
   // likewise for SHA-3

   default:
      throw Not_Implemented("SPHINCS+ is not implemented for this hash type");
}
```

So that the function always either returns a value or throws.

Reply:
My two cents: If we shy away from `unreachable()` we should at least move the `Not_Implemented` outside and under the entire switch statement. Omitting the `default` case in enum-based switches enables useful compiler warnings when the enum is amended but the switch is forgotten.